core/mo-services-impl/nmf-platform-generic-impl/src/main/java/esa/mo/platform/impl/provider/gen/GPSProviderServiceImpl.java

```diff
@@ -483,63 +518,95 @@ public GetLastKnownPositionAndVelocityResponse getLastKnownPositionAndVelocity(
+          Vector3D pos = state.getPVCoordinates().getPosition();
+          position = new VectorD3D(pos.getX(), pos.getY(), pos.getZ());
 
-    } catch (IOException | NumberFormatException e) {
-      interaction
-          .sendError(new MALStandardError(PlatformHelper.DEVICE_NOT_AVAILABLE_ERROR_NUMBER, null));
-      e.printStackTrace();
-    }
+          positionDeviation = new VectorF3D(0f, 0f, 0f);
+
+          Vector3D velocity3D = state.getPVCoordinates().getVelocity();
+          velocity = new VectorD3D(velocity3D.getX(), velocity3D.getY(), velocity3D.getZ());
+
+          velocityDeviation = new VectorF3D(0f, 0f, 0f);
+        }
+        else {
+          String bestxyz = adapter.getBestXYZSentence();
+
+          String[] fields = HelperGPS.getDataFieldsFromBestXYZ(bestxyz);
+
+          position = new VectorD3D(
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.PX]),
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.PY]),
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.PZ])
+          );
+
+          positionDeviation = new VectorF3D(
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.PX_DEVIATION]),
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.PY_DEVIATION]),
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.PZ_DEVIATION])
+          );
+
+          velocity = new VectorD3D(
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.VX]),
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.VY]),
+                  Double.parseDouble(fields[HelperGPS.BESTXYZ_FIELD.VZ])
+          );
+
+          velocityDeviation = new VectorF3D(
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.VX_DEVIATION]),
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.VY_DEVIATION]),
+                  Float.parseFloat(fields[HelperGPS.BESTXYZ_FIELD.VZ_DEVIATION])
+          );
+        }
+
+        synchronized (MUTEX) { // Store the latest Position
+          currentCartesianPosition = position;
+          currentCartesianPositionDeviation = positionDeviation;
+          currentCartesianVelocity = velocity;
+          currentCartesianVelocityDeviation = velocityDeviation;
+          timeOfCurrentPositionAndVelocity = System.currentTimeMillis();
+        }
+
+        interaction.sendResponse(position, positionDeviation, velocity, velocityDeviation);
+
+      } catch (IOException | NumberFormatException e) {
+        interaction
+                .sendError(new MALStandardError(PlatformHelper.DEVICE_NOT_AVAILABLE_ERROR_NUMBER, null));
+        e.printStackTrace();
+      }
 
   }
 
   @Override
   public void getTLE(GetTLEInteraction interaction) throws MALInteractionException, MALException
   {
-    if (!adapter.isUnitAvailable()) { // Is the unit available?
+    if (!adapter.isUnitAvailable() && isTLEFallbackEnabled == false) { // Is the unit available?
```

Comment:
Or should TLE be always available? On OPS-SAT this is read from a file, so should not pose any issues.